src/hydrosheds.cc

```diff
@@ -6,96 +6,306 @@
+
+        return Coordinate{end.getX(), end.getY()};
+    }
+
+    bool RiverSegment::hasDownstreamSegment() const
+    {
+        // If the NEXT_DOWN field is 0, there is no downstream segment
+        if (feature->GetFieldAsInteger("NEXT_DOWN") == 0)
+            return false;
+
+        // If the ENDORHEIC field is 1, there is also no downstream segment
+        // TODO: This is my limited understanding - might need to rethink
+        if (feature->GetFieldAsInteger("ENDORHEIC") == 1)
+            return false;
+
+        // In all other cases, there is a downstream segment;
+        return true;
+    }
+
+    double RiverSegment::getLength() const
+    {
+        return getLineString(feature)->get_Length();
+    }
+
+    double RiverSegment::getGeologicalLength() const
+    {
+        return feature->GetFieldAsDouble("LENGTH_KM");
+    }
+
+    // begin and end of the hydroshedsdata
+    FullDatasetRiverSegmentIterator HydroshedsDataSet::begin() const
+    {
+        this->layer->ResetReading();
+        auto feature = layer->GetNextFeature();
+        FullDatasetRiverSegmentIterator start(feature, layer);
+        return start;
+    }
+
+    FullDatasetRiverSegmentIterator HydroshedsDataSet::end() const
+    {
+        FullDatasetRiverSegmentIterator end(NULL, layer);
+        return end;
+    }
+
+
+    // FullDataSetRiverSegmentIterator
+    FullDatasetRiverSegmentIterator::FullDatasetRiverSegmentIterator(OGRFeature* ogrFeature, OGRLayer* ogrLayer)
+            : segment(ogrLayer, ogrFeature)
+    {
+        // intializing the values
+        feature = ogrFeature;
+        layer = ogrLayer;
+    }
+
+
+
+
+    // Prefix increment
+    FullDatasetRiverSegmentIterator FullDatasetRiverSegmentIterator::operator++()
+    {
+        // "incrementing" the Feature first
+        //FullDatasetRiverSegmentIterator t(feature, layer);
+        auto nextFeature = this->layer->GetNextFeature();
+        this->segment.feature = nextFeature;
+        this->feature = nextFeature;
+        return (*this);
+    }
+
+    // Postfix increment
+    FullDatasetRiverSegmentIterator FullDatasetRiverSegmentIterator::operator++(int)
+    {
+        // returning the feature first
+        auto result = (*this);
+        auto nextFeature = this->layer->GetNextFeature();
+        this->segment.feature = nextFeature;
+        this->feature = nextFeature;
+        return result;
+    }
+
+    // Two implementations for the comparison operators
+
+    bool FullDatasetRiverSegmentIterator::operator==(const FullDatasetRiverSegmentIterator& a)
+    {
+        bool result;
+        //result = feature->Equal(a.feature);
+        result = feature == a.feature;
+        return result; // do pointer equality
+    }
+
+
+    bool FullDatasetRiverSegmentIterator::operator!=(const FullDatasetRiverSegmentIterator& a)
+    {
+        bool result;
+        //result = feature->Equal(a.feature);
+        result = feature == a.feature;
+        return !result; // do pointer equality
```

Comment:
A conventient way of implementing `operator!=` that does not duplicate logic is to call `!operator==`.